### DISCUSSION TITLE
The navigation results are confusing me

### DISCUSSION THREAD

**Discussion by ytipocmk631** [recastnavigation/recastnavigation#732] [Q&A]
Hello, this project is great and has solved many of my problems. However, in some places (as shown in the picture), the navigation results generated have confused me. How can I modify this problem?
![image](https://github.com/user-attachments/assets/dfb3ea05-c87a-4962-b106-94dd40ceae32)

**Comment by stynosen** [comment]
Hi!

I saw the postmortem of Death Stranding a while back, and they mention this exact issue.
https://youtu.be/yqZE5O8VPAU?si=tJ5S3LRDQJby85Ev&t=1156

They reference Miko's blog as well with a potential solution
https://digestingduck.blogspot.com/2010/03/simple-stupid-funnel-algorithm.html 

Hope that helps you!

**Comment by InspAlgo** [comment]
You can add raycast to detect, for example https://github.com/recast4j/recast4j/blob/49a44880293cbc0034e61b2118cfe260069d1b6f/detour/src/main/java/org/recast4j/detour/NavMeshQuery.java#L812
![image](https://github.com/user-attachments/assets/a4c0f417-aba4-4a5f-a037-12979b36bfd1)

**Reply by sagaceilo** [reply]
It's interesting that 

**dtNavMeshQuery::updateSlicedFindPath** is already using tryLOS but with different check. Probably just adding this would do: 


https://github.com/recastnavigation/recastnavigation/blob/455a019e7aef99354ac3020f04c1fe3541aa4d19/Detour/Source/DetourNavMeshQuery.cpp#L1410
`foundShortCut = rayHit.t >= 1.0f && rayHit.path[rayHit.pathCount-1] == neighbourRef;`

but **dtNavMeshQuery::findPath** is not.

https://github.com/recastnavigation/recastnavigation/blob/455a019e7aef99354ac3020f04c1fe3541aa4d19/Detour/Source/DetourNavMeshQuery.cpp#L1090

Does it mean that its partially implemented on master branch of recast query ? Any suggestions ?

**Reply by InspAlgo** [reply]
Using `foundShortCut = rayHit.result.t >= 1.0f && rayHit.result.path.get(rayHit.result.path.size() - 1) == neighbourRef;` instead of `foundShortCut = rayHit.result.t >= 1.0f;` can refer to https://github.com/recast4j/recast4j/pull/197.

As for why it wasn’t used in `findPath`, I’m not entirely sure, but based on my experience, the reasons are probably as follows: 
1. The terrain can be modified. 
2. The build parameters can be adjusted (perhaps `Tile` is more suitable than `Solo`). 
3. Adding `raycast` for correction in `findPath` would negatively affect performance, leading to longer function runtime. 
 
The performance issue might be key, because in game movement handling, `findPath` is likely called frequently, making it a critical point for performance optimization.

**Reply by sagaceilo** [reply]
Thanks for clarification!
